code/GatewayFieldsFactory.php

```diff
@@ -320,47 +310,133 @@ protected function cullForGateway(&$fields, $defaults = array())
      *
      * @return array
      */
-    public function normalizeFormData(array $data) {
+    public function normalizeFormData(array $data)
+    {
+        $renameMap = $this->getRenameMap();
+
+        $prefix = $this->getGatewayPrefix();
+        // if a prefix has already been defined in the gateway namespace, continue using it
+        // if not, and we have a global prefix, use that.
+        // if not, $prefix becomes an empty string
+        $prefix = ($prefix) ? $prefix : ($this->getGlobalPrefix() ?: '');
+
+        // strip prefix first
+        foreach ($data as $key => $value) {
+            if (strstr($key, $prefix)) {
```

Comment:
I think it would be better to check if the prefix is at the beginning. With the current implementation you might rename a field `mytest_data` to `mydata` if your prefix is `test_`. So maybe use `strpos($key, $prefix) === 0` instead?